### ISSUE TITLE
force_deregister does not clean up old snapshots

### ISSUE THREAD

**Issue by eedwardsdisco** [mitchellh/packer#3320]
Packer v0.9.0

When setting 'force_deregister', Packer will de-register the AMI but will not clean up the snapshots associated with the AMI.

force_deregister should either also cleanup snapshots, or another option should be provided that enables that behavior

**Comment by b-dean** [comment]
FWIW, deregister doesn't delete snapshots in the [AWS Console](https://console.aws.amazon.com) either. I suspect the `aws` cli doesn't either: http://docs.aws.amazon.com/cli/latest/reference/ec2/deregister-image.html

> This command does not delete the AMI.

I'm guessing that means it doesn't delete the snapshots.

**Comment by eedwardsdisco** [comment]
Like I said, an option to enable that behavior would be nice.

**Comment by Constantin07** [comment]
Completely agree with @eedwardsdisco

Just faced this issue too - it doesn't nicely clean up. Yes, indeed AWS doesn't do that but once ssh key pair and security group are temporarily created by packer for the AMI creation and then deleted, it would be consistent to delete old snapshot too.
Now it looks like a half solution - something done automatically and something left to be carried out manually.

**Comment by mwhooker** [comment]
according to the docs

```
force_deregister (boolean) - Force Packer to first deregister an existing AMI if one with the same name already exists.
```

force_deregister is only used to clean up existing AMIs if there would be a conflict, so I think its current behavior is correct.

Are you just saying that you don't want any snapshots lying around after the build? I'm not sure that's possible since amazon needs a snapshot of the root ebs device to launch the AMI.

Or do you just want old snapshots cleaned up when force_deregister is true?

**Comment by mwhooker** [comment]
on a re-reading of the issue I think you're asking for the latter, to clean old snapshots when we clean old images. I think that makes total sense. please reopen if I got it wrong

**Comment by eedwardsdisco** [comment]
@mwhooker sorry, been busy at AWS re:Invent (even stopped by the hashicorp booth to get some swag), but yes you're 100% right, looking for the ability to clean up associated snapshots when deregistering the image.

As @Constantin07 said, the behavior of leaving the snapshots after deregistering the AMIs just doesn't make sense if packer is generating the snapshots in the first place and they're useless after the fact (e.g. if I keep re-generating an AMI using force_deregister and it just keeps leaving lingering snapshots that I'm not using).

Basically, ideally, when it goes to de-register the old AMI, it also removes snapshots.

Really appreciate the work on supporting this use case :)

**Comment by eedwardsdisco** [comment]
Sorry I re-read your question and realized I may have made it more confusing.

What's not desired is that when Packer deregisters an AMI due to a conflict, it leaves the snapshots behind.

**Comment by mwhooker** [comment]
okay, I think we've got the correct behavior in #4223 now.

Hope you got some good swag!
